### DISCUSSION TITLE
Smart Meter Script for S0 Interface

### DISCUSSION THREAD

**Discussion by TBR-BRD** [arendst/Tasmota#16294] [Q&A] [answered]
Is there an example for the integration of a meter with S0 interface? https://xn--stromzhler-v5a.eu/stromzaehler/wechselstromzaehler/fuer-hutschiene-ungeeicht/66/sdm120-pls-1-ph.-wsz-fuer-din-hutschiene
It outputs 1000 pulses per KWh.

**Comment by Ekran** [comment, marked as answer]
Hello, do you have a working example in the meantime? Thanks!

**Reply by sfromis** [reply]
Tasmota can handle pulses by configuring a pin as a counter.

**Reply by TBR-BRD** [reply]
Hello, no.

**Reply by Ekran** [reply]
from https://tasmota.github.io/docs/Smart-Meter-Interface/ you find the settings for Main Menu --> Consoles --> Edit Script

for a S0 impuls contact with 100 pulses per kWh:

```
>D
>B
>M 1
+1,2,c,1,-10,Energy
1,1-0:1.8.0*255(@100,consumption,KWh,Total_in,3)
1,1-0:1.7.0*255(@0.0016667, power,W,Power_actual,0)
;
```

and for 1000 pulses per kWh:
```
>D
>B
>M 1
+1,2,c,1,-10,Energy
1,1-0:1.8.0*255(@1000,consumption,KWh,Total_in,3)
1,1-0:1.7.0*255(@0.01667, power,W,Power_actual,0)
#
```

**Reply by Elektrofix-OL** [reply]
The documentation says that 5 meters can be connected, but is it possible that only 2 couter are possible?
When I try 4 counters with S0 in the script, the ESP crashes and when I reboot, script enabled is off.

**Reply by gemu2015** [reply]
please post your script, i will check what is going on.  MAX_COUNTERS is predefined to 4, so 4 counters should be supported.

**Reply by Elektrofix-OL** [reply]
Here the script. Tried with ESP8266 with 4 and 16Mbyte memory

```
>D
>B
=>sensor53 r
>W
;<style>form button{display:none}</style>
>M 4
+1,6,c,1,-10,Garten
1,1-0:1.8.0*255(@1000,consumption,KWh,Total_in,3)
1,1-0:1.7.0*255(@0.01667, power,W,Power_actual,0)
+2,7,c,1,-10,Strasse
2,1-0:1.8.0*255(@1000,consumption,KWh,Total_in,3)
2,1-0:1.7.0*255(@0.01667, power,W,Power_actual,0)
+3,2,c,1,-10,Bahn
3,1-0:1.8.0*255(@1000,consumption,KWh,Total_in,3)
3,1-0:1.7.0*255(@0.01667, power,W,Power_actual,0)
+4,0,c,1,-10,Dach
4,1-0:1.8.0*255(@1000,consumption,KWh,Total_in,3)
4,1-0:1.7.0*255(@0.01667, power,W,Power_actual,0)
#
```

**Reply by gemu2015** [reply]
ok you did choose "forbidden" pins for an ESP8266
at least with "allowed" pins it does not crash.
you may choose e.g. pins 12,13,14,15

**Reply by steand** [reply]
pin 15 (D8) isn't a good choice. D8 has a pulldown an must LOW at boot. If a external pullup exist the ESP2866 didn't boot.
Better use: 4,12,13,14.

This works for me:

```
>D
>B
=>sensor53 r
>M 4
+1,12,c,1,-10,Privat
1,1-0:1.8.0*255(@400,Bezug,KWh,Total_in,3)
+2,13,c,1,-10,Mietw
2,1-0:1.8.0*255(@400,Bezug,KWh,Total_in,3)
+3,14,c,1,-10,Keller
3,1-0:1.8.0*255(@400,Bezug,KWh,Total_in,3)
+4,4,c,1,-10,Allg
4,1-0:1.8.0*255(@400,Bezug,KWh,Total_in,3)

```

**Comment by mafde2023** [comment]
I want to use that script to monitor my heat pump energy consumption which has an S0 meter installed before
circuit breakers. Is there any instruction how to do the wiring at ESP8266/Wemos D1 board with Tasmota?

**Comment by Juhe0167** [comment]
Hello,
I will also control the power consumption from my heat  pump.
I use an ESP Wroom32 with Tasmota 13.2.0, an created a counter on GPIO15.
It is counting and I can see it on Grafana, so MQTT works.

I use this ESP32 and Tasmota allready for measurement with DS18B20-sensors and it's working fine.

I'm sorry, but I do not understand how to get the rest running.
When I try to copy the Scipts in this thread into the Berry Console, I always get an error message, f.e.:
 >D
 >B
 =>sensor53 r
 >W
 ;<style>form button{display:none}</style>
 >M 1
 +1,14,c,1,-40,WP
 1,1-0:1.8.0*255(@1000,consumption,KWh,Total_in,3)
 1,1-0:1.7.0*255(@0.01667, power,W,Power_actual,0)
 #

BRY: Exception> 'syntax_error' - input:1: unexpected symbol near '>'

Sorry, the > is not typed here, but in the scipt...

Whats wrong?

I've read a lot, but do not get any advantage.

Maybe, someone can help me or give me a working scipt.

**Reply by sfromis** [reply]
Berry is a very different language; what you have there is for the older scripting feature, which is only available in the `tasmota32-scripting` found here: https://github.com/Jason2866/Tasmota-specials/tree/firmware/firmware/tasmota32/other
(Also available in the Tasmota web installer under "unofficial" binaries).

**Reply by Juhe0167** [reply]
That's it, I installed the ESP new with the scripting-version an now it is working.
Thanks a lot.
